compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -741,25 +729,374 @@ fn encode_ty<'tcx>(
+                    ),
+                )
+            }
+        };
+        let mut fn_abi = fn_abi.clone();
+        // HACK(rcvalle): It is okay to not replace or update the entire ArgAbi here because the
+        //   other fields are never used.
+        fn_abi.args[0].layout.ty = self_ty;
+        return fn_abi;
+    } else if let Some(instance) = instance
+        && matches!(instance.def, ty::InstanceDef::VTableShim(..))
+    {
+        // Adjust the type ids of VTableShims
+        return transform_closure_fnabi(tcx, fn_abi, options);
+    } else if let Some(instance) = instance
+        && matches!(instance.def, ty::InstanceDef::DropGlue(..))
+    {
+        // Adjust the type ids of DropGlues
+        if !fn_abi.args.is_empty() {
+            let self_ty = if fn_abi.args[0].layout.ty.is_mutable_ptr() {
+                Ty::new_mut_ref(
+                    tcx,
+                    tcx.lifetimes.re_erased,
+                    new_dynamic_trait(tcx, tcx.lang_items().drop_trait().unwrap(), List::empty()),
+                )
+            } else {
+                Ty::new_imm_ref(
+                    tcx,
+                    tcx.lifetimes.re_erased,
+                    new_dynamic_trait(tcx, tcx.lang_items().drop_trait().unwrap(), List::empty()),
+                )
+            };
+            let mut fn_abi = fn_abi.clone();
+            // HACK(rcvalle): It is okay to not replace or update the entire ArgAbi here because the
+            //   other fields are never used.
+            fn_abi.args[0].layout.ty = self_ty;
+            return fn_abi;
+        }
+        return fn_abi.clone();
+    } else if let Some(instance) = instance
+        && matches!(instance.def, ty::InstanceDef::ClosureOnceShim { .. })
+    {
+        // Adjust the type ids of ClosureOnceShims
+        return transform_closure_fnabi(tcx, fn_abi, options);
+    } else if is_closure_fnabi(tcx, fn_abi) {
+        // Transform closure FnAbis into funtion pointer FnAbis
+        return transform_closure_fnabi(tcx, fn_abi, options);
+    } else if is_dynamic_fn_fnabi(tcx, fn_abi) {
+        // Transform Fn trait FnAbis into function pointers FnAbis
+        return transform_fn_trait_fnabi(tcx, fn_abi, options);
+    }
+    return fn_abi.clone();
+}
+
+/// Transforms a predicate for being encoded and used in the substitution dictionary.
+fn transform_predicate<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    predicate: ty::PolyExistentialPredicate<'tcx>,
+    _options: EncodeTyOptions,
+) -> ty::PolyExistentialPredicate<'tcx> {
+    match predicate.as_ref().skip_binder() {
+        ty::ExistentialPredicate::Trait(trait_ref) => {
+            if tcx.is_fn_trait(trait_ref.def_id) {
+                // Fn trait objects are transformed into function pointers in transform_ty
+                bug!("transform_predicate: unexpected Fn trait `{}`", trait_ref)
+            } else {
+                // Transform non Fn trait objects into their identities
```

Comment:
I don't think this makes sense. Please consider another approach that doesn't require projecting trait objects to their *identity args*, but if this is necessary, then like I said above with drop shims, this needs to be motivated by a code example that we want to pass CFI.

Reply:
This is also related to the information I provided above. Since we can can only include at most as much information that would be available at the call sites, otherwise the labels won't match. The maximum amount of information we can include at the call site for calls on trait objects we can obtain is by resolving the trait identity of the trait that implements the method (i.e., the unsubstituted/unmonomorphized trait/the trait as it is declared), we perform type erasure on both calls to `typeid_for_fnabi` to trait identity of the trait that implements the method, so the labels match.